src/lua/protobuf.lua

```diff
@@ -657,9 +726,142 @@ end
 
 -- }}} Encoders
 
+-- {{{ Decoders
+
+local function read_unknown_field(ctx, type)
+    local res = string.sub(ctx.val, ctx.pos - 1, ctx.pos - 1)
+    local begin = ctx.pos
+    if type == WIRE_TYPE_LEN then
+        local len = wireformat.decode_int(ctx)
+        ctx.pos = ctx.pos + len
+    elseif type == WIRE_TYPE_I32 then
+        ctx.pos = ctx.pos + 4
+    elseif type == WIRE_TYPE_I64 then
+        ctx.pos = ctx.pos + 8
+    else
+        local _ = wireformat.decode_int(ctx)
+    end
+    res = res .. string.sub(ctx.val, begin, ctx.pos - 1)
+    return res
+end
+
+local function decode_repeated(protocol, field_def, ctx)
+    local res = {}
+    if ctx.packed then
+        local len = wireformat.decode_int(ctx)
+        local start_pos = ctx.pos
+        local value
+        while ctx.pos < start_pos + len do
+            ctx.ignore_repeated = true
+            value = decode_field(protocol, field_def, ctx)
+            table.insert(res, value)
+        end
+    else
+        repeat
+            local field_id
+        ctx.ignore_repeated = true
+            local value = decode_field(protocol, field_def, ctx)
+            table.insert(res, value)
+            if ctx.pos > string.len(ctx.val) then
+                return res
```

Comment:
Shouldn't we compare with some amount-of-elements in the message or size-in-bytes of the message? The global position check seems irrelevant at parsing of the particular nested message (except the buffer boundary check at reading a byte).

Reply:
It seems like we it is the end of a source data, not the end of a substring.

The approach with extracting a substring with the source data of a nested message feels confusing for me. This way we need to re-create the decoding context, and it also seems not like contexts usually work.

We can accept a source data not as just a string, but the string + its length. This way we eliminate extracting a substring for each nested message and eliminate a need to re-create the context.

The leght should be passed in arguments, together with data (we can't store it in context, a per-frame value is needed).

While a string extraction approach reduces amount of arguments we need to pass over the functions (no explicit length), I think it worth to consider my proposal. At least, it was a surprise for me that the decoding context is re-created and I misinterpreted this check.

Reply:
In context of message decoding this seems strange for me. Let's imagine that in the middle of decoding we decode field_id and type and then we decide that the next one field is message type and decoded as `LEN`. It means that we can read known amount of bytes and it contains full embedded message. We add existing path to new context and start decoding embedded message and guarantee that we will not run out of embedded message and add extra fields in it.

If we will not create new decoding context we have to pass length of embedded message in `decode_message` function and watch until position runs out of this boundary. As I see it can lead to unexpected behaviour. For example we have a protocol describing this message:
```
protocol = proto.protocol({
    proto.message('test_message', {
        val1 = {'int32', 1},
        val2 = {'test_message', 2},
        val3 = {'int32', 20000},
    })
})
```
And somebody encoded this values:
```
protocol:encode('test_message', {val1 = 12, val2 = {val1 = 500, val3 = 1}, val3 = 25})
```

As a result we will get this binary string: `08 0c 12 07 80 e2 09 01 08 f4 03 80 e2 09 19` where
`08 0c` - val1 =12
`12 07 80 e2 09 01 08 f4 03` - val2, where `80 e2 09 01` - val3 = 1, `08 f4 03` - val1 = 500
`80 e2 09 19` - val3 = 25

Lets say we lost one byte `03` then we will get this string: `08 0c 12 07 80 e2 09 01 08 f4 <03> 80 e2 09 19`
Now my decoder will take embedded message `12 07 80 e2 09 01 08 f4 80` as a new decoding context and return such an error: ` While decode first_message.val2.val3 on byte 7 occurred error: Reached end of message while decoding'

But if we will not create a new decoding context the decoder can read `08 f4 80 e2 09` because msb will be zero only on `09` byte. And wireformat function will not detect reading out of bound. Yes we can detect it later when we will return to `decode` function but how can we say on which byte the error occured?

Reply:
Thanks for the example!

Let's look into it from the single-context-and-length-argument approach perspective.

> Lets say we lost one byte `03` then we will get this string: `08 0c 12 07 80 e2 09 01 08 f4 <03> 80 e2 09 19`
>
> Now my decoder will take embedded message `12 07 80 e2 09 01 08 f4 80` as a new decoding context and return such an error: <...>

I'll sketchy picture a call tree.

* decode_outer_message(s: `08 0c 12 07 80 e2 09 01 08 f4 <03> 80 e2 09 19`, len: 14)
  * decode_int32(s: `08 0c`, len: 2)
  * decode_nested_message(s: `12 07 80 e2 09 01 08 f4 <03> 80`, len: 9)
    * decode_int32(s: `80 e2 09`, len: 3)
    * decode_int32(s: `08 f4 <03> 80`, len: 3)
      => boundary error

We can't continue reading bytes beyond the given length, so we get the same buffer boundary error.

In fact, both approaches are the same from the algorithmic point of view.

However, I have two reasons to consider the single-context-and-length-argument approach:

1. Imagine we write this code on C. We likely keep some pointer from-where-to-read-next-byte and a right boundary *for the given call frame*. We unlikely allocate a new buffer and copy a part of the source string on each call frame.

   Now we're in Lua and it is a bit non-trivial to eliminate copying. It is OK. However, I would like if the code is written in a way similar to how we write it in C. This way we can re-write parts using FFI or C without major changes in the code.
2. More important point is how another person recognizes typical programming patterns from your code.

   I see a resursive algorithm that pass some context over the functions. Usually a context is needed in recursive code to keep something across the call frames. If something is named as 'context' or 'ctx', I can't ever imagine that it is new for each frame. It is not a context, it is something completely different. Maybe 'frame'. Or, at least 'message-local-context'.

   If we need to pass frame-related information in a recursive algorithm, we have two straightforward ways:

   * Pass it as an argument.
   * Use a stack.

   Using a partially-copied partially-global-but-partially-frame-local, hmm, context is not a direct way to do so. My pattern recognition fails here.

Reply:
I got your point, now I pass bounds for decoding as a function parameter and context stays unchanged. I added mentioned above testcase into tests.